### PR TITLE
Bug fix: Limiting Number of Sponsorship Calls for New Users

### PR DESCRIPTION
**Pull request by lykimq**
**Git Issue #45: Limiting Number of Sponsorship Calls for New Users**

**Description:**
Currently, there exists a manual limitation on the number of calls for a specific sponsee (giving their address and so on).

However, the correct condition should be to set a maximum limit on a number of calls from any address or new user, without having to manually enter them one by one.  This ensures that only the first N calls are sponsored for each user.

- Remove the current `sponsee_address` in condition
- Add the option to choose the contract to set a condition
- Add a condition for N new users (tie to the chosen contract), currently the condition for the users is the last 4 newest users registered in the database.

### CODE REVIEW

File: src/crud.py [hunk 1]
```diff
@@ -317,26 +317,30 @@ def check_calls_per_month(db, contract_id):
         db, contract_id)
     return max_calls >= len(nb_operations_already_made)
 
+# Function to limit the number of calls for new users
+# rather than for a specific sponsee.
+
```
Comment (aguillon) [inline src/crud.py]: Please use docstrings to document functions instead, when it makes sense

Comment (lykimq) [reply src/crud.py]: put it in the doc string at this [commit](https://github.com/marigold-dev/gas-station/pull/52/commits/d7063083fea97074120ca4309d857da92149491a)

File: src/crud.py [hunk 2]
```diff
@@ -415,20 +427,21 @@ def check_max_calls_per_entrypoint(
 
 def check_conditions(db: Session, datas: schemas.CheckConditions):
     print(datas)
-    sponsee_condition = check_max_calls_per_sponsee(
-        db, datas.sponsee_address, datas.vault_id
-    )
+
+    max_calls_condition = check_max_calls_per_sponsee
+    (db, datas.vault_id)
+
```
Comment (aguillon) [inline src/crud.py]: This is valid Python, but does not do what you think it does. If your code formatter produced this, then it is broken.

File: src/crud.py [hunk 3]
```diff
@@ -438,8 +451,8 @@ def check_conditions(db: Session, datas: schemas.CheckConditions):
     # Update conditions
     # TODO - Rewrite with list
 
-    if sponsee_condition:
-        update_condition(db, sponsee_condition)
+    if max_calls_condition:
+        update_condition(db, max_calls_condition)  # type: ignore
```
Comment (aguillon) [inline src/crud.py]: Can you explain why you had to add this `type: ignore` annotation?

File: src/models.py [hunk 4]
```diff
@@ -117,53 +117,51 @@ class Operation(Base):
...
 
+    # Add a CheckConstraint to ensure consistency between type and associated fields
+    __table_args__ = (
+        CheckConstraint(
+            "(type = 'MAX_CALLS_PER_ENTRYPOINT' AND entrypoint_id IS NOT NULL) \
+          OR (type = 'MAX_CALLS_PER_ENTRYPOINT') AND contract_id IS NOT NULL)",
+            name="type_to_foreign_key_constraint",
+        ),
+    )
+
```
Comment (aguillon) [inline src/models.py]: Why do you replace two consecutive constraints by a OR? This does not make sense to me.

Also, is there no constraint on max sponsee condition?

File: src/routes.py [hunk 5]
```diff
@@ -268,8 +272,11 @@ async def get_entrypoint(
 # Operations
 @router.post("/operation")
 async def post_operation(
+    # call_data: contains information about the operations to be performed
```
Comment (aguillon) [inline src/routes.py]: You should use a docstring here too

File: src/routes.py [hunk 6]
```diff
@@ -299,14 +308,19 @@ async def post_operation(
...
             if not entrypoint.is_enabled:
                 raise EntrypointDisabled()
 
+            # Check if certain conditions are met for each operation,
+            # if not, raises corresponding exceptions
             if not crud.check_conditions(
                 db,
                 schemas.CheckConditions(
-                    sponsee_address=call_data.sender_address,
+                    sponsee_address=None,  # Remove sponsee address
```
Comment (aguillon) [inline src/routes.py]: What's the point of keeping the argument if you're going to set it to None?

File: src/schemas.py [hunk 7]
```diff
@@ -154,13 +154,12 @@ class CreateMaxCallsPerEntrypointCondition(BaseModel):
...
 
 class CreateMaxCallsPerSponseeCondition(BaseModel):
-    sponsee_address: str
     vault_id: UUID4
     max: int
 
 
 class CheckConditions(BaseModel):
-    sponsee_address: str
+    sponsee_address: Optional[str]
```
Comment (aguillon) [inline src/schemas.py]: Same observation as before

File: src/schemas.py [hunk 8]
```diff
@@ -181,4 +180,4 @@ class MaxCallsPerEntrypointCondition(ConditionBase):
 
 
 class MaxCallsPerSponseeCondition(ConditionBase):
-    sponsee_address: str
+    pass
```
Comment (aguillon) [inline src/schemas.py]: Who does this condition apply to?

**Review by aguillon — CHANGES_REQUESTED** [review]
I haven't tried due to mysterious, but hopefully temporary, technical reasons. However, there's no way this code works. Can you please describe how you tested it?

File: src/models.py (hunk 4 above)
Comment (lykimq) [reply src/models.py]: I reverse it back as before as this [commit](https://github.com/marigold-dev/gas-station/pull/52/commits/2cc8de0bcabc7c6775a160310e3be478ebf7d1e5), 
There is no constraint for max sponsee condition because it does not depend on the sponsee_address, but it is 
at these 2 fields:
```
    # New field to store the maximum allowed calls per user
    max_calls_per_user = Column(Integer, nullable=False)
    # New field to track current calls per user
    current_calls_per_user = Column(Integer, nullable=False, default=0)
```

File: src/schemas.py (hunk 8 above)
Comment (lykimq) [reply src/schemas.py]: It was called from this function [create_max_calls_per_sponsee_condition](https://github.com/marigold-dev/gas-station/blob/quyen%40max_sponsee/src/crud.py#L350)

File: src/crud.py (hunk 2 above)
Comment (lykimq) [reply src/crud.py]: You are right, it was the formatter broken, I correct it at this [commit](https://github.com/marigold-dev/gas-station/pull/52/commits/9e0df0e32096cfa6b78897aa5bfbc002dfafbad7#diff-b2e3e31168d8f83b0946e8e58999dac2a49bf9087a3e20526f942694bb2a0e30R431)

File: src/crud.py (hunk 3 above)
Comment (lykimq) [reply src/crud.py]: I remove the `type: ignore`

File: src/crud.py (hunk 2 above)
Comment (lykimq) [reply src/crud.py]: I wasn't able to test it yet.

File: src/routes.py (hunk 5 above)
Comment (lykimq) [reply src/routes.py]: I have remove it.

File: src/schemas.py (hunk 8 above)
Comment (lykimq) [reply src/schemas.py]: It has been rework, so I resolved it.

File: src/schemas.py (hunk 7 above)
Comment (lykimq) [reply src/schemas.py]: It has been reword.

File: src/schemas.py [hunk 9]
```diff
@@ -180,5 +181,6 @@ class MaxCallsPerEntrypointCondition(ConditionBase):
     entrypoint_id: UUID4
 
 
-class MaxCallsPerSponseeCondition(ConditionBase):
-    sponsee_address: str
+class MaxCallsPerContractForNewUsersCondition(ConditionBase):
+    contract_id: UUID4
+    user_id: UUID4
```
Comment (aguillon) [inline src/schemas.py:186]: Why did you add this?

File: src/crud.py [hunk 10]
```diff
@@ -44,6 +44,26 @@ def create_user(db: Session, user: schemas.UserCreation):
...
+    """
+    Function to determine if a user is one of the last 4
+    new users registered in the database.
+    """
+    try:
+        # Get the last 4 users ordered by their registration data
+        recent_users: List[models.User] = (
+            db.query(models.User)
+            .order_by(models.User.created_at.desc())
+            .limit(4)
```
Comment (aguillon) [inline src/crud.py:57]: What? Why such an arbitrary limit?

File: src/routes.py [hunk 11]
```diff
@@ -299,14 +303,24 @@ async def post_operation(
         entrypoint_name = operation["parameters"]["entrypoint"]
 
         try:
-            entrypoint = crud.get_entrypoint(db, str(contract.address), entrypoint_name)
+            entrypoint = crud.get_entrypoint(
+                db, str(contract.address), entrypoint_name)
             if not entrypoint.is_enabled:
                 raise EntrypointDisabled()
 
+            # Retrieving the user_id
```
Comment (aguillon) [inline src/routes.py:311]: In this context, it's not clear who you call "the user". According to what you write just after, it seems to be the user who added the contract; however, you're also testing if the user is "new" (for some definition of "new", but whatever), which, I suspect, is your way of translating the real-world example I gave you. In this example, I was giving you the idea that we needed to change the "max calls per user" condition so that it can be used to limit the activity of some users without impacting newer users, but your translation of this requirement seems rather fanciful.

Comment (aguillon) [reply src/routes.py:311]: The general idea is that, in the context of this issue/PR, a user should not be considered "new" because of the age of their account, but rather because of their recent activity relatively to the sponsored contract.

However, when we talk about "user" in this context, we're definitely talking about end-users, or sponsees (users whose transactions are paid for by the API) and not contracts' owners.

File: src/routes.py [hunk 12]
```diff
@@ -299,14 +303,24 @@ async def post_operation(
...
+            # Check that if the user_id is among the 4 newest user
+            owner_address = contract.owner.address
+            user = crud.get_user_by_address(db, owner_address)
+            user_id = user.id
+            if not crud.check_user_is_new(db, user_id.id):
+                raise UserNotFound()
```
Comment (aguillon) [inline src/routes.py:317]: So… if the user is not new, you're simply refusing the operation, regardless of the condition?

**Review by aguillon — CHANGES_REQUESTED** [review]
This does not seem correct at all, for various reasons.
